### PR TITLE
144 connect picker with editing functionality

### PR DESCRIPTION
**Pull request by lydiaklBal**
Added the functionality to update the position and rotation of objects and implemented a check to determine whether an object is movable or rotatable. In addition fixed minor bugs.

### CODE REVIEW

File: canvas_project/editor/templates/editor/editor.html [hunk 1]
```diff
@@ -308,7 +308,7 @@ <h5 class="modal-title" id="aboutModalLabel">About</h5>
             </div>
             <!--quick settings-->
             <div class="position-absolute bottom-0 mb-4 fs-5 d-flex justify-content-center w-100">
-                <div class="p-2 gap-2 shadow-sm bg-body-secondary bg-opacity-75 rounded-4 border d-flex"
+                <div class="p-2 gap-2 shadow-sm bg-body-secondary bg-opacity-75 rounded-4 border d-flex" id="pillNav1"
```
Comment (plhrtr) [inline canvas_project/editor/templates/editor/editor.html:311]: I think the id should describe the object

File: canvas_project/static/js/projectSettingsManager.mjs [hunk 2]
```diff
@@ -108,13 +108,11 @@ export class ProjectSettingsManager {
         checkbox.type = "checkbox";
         checkbox.classList.add("form-check-input");
         checkbox.checked = isChecked;
-        checkbox.id = label + "Checkbox";
```
Comment (plhrtr) [inline canvas_project/static/js/projectSettingsManager.mjs:111]: does lines shouldnt be removed, as they allow to select the check box via the text and not only via the box

File: canvas_project/static/js/projectSettingsManager.mjs [hunk 3]
```diff
@@ -108,13 +108,11 @@ export class ProjectSettingsManager {
...
 
         //label for the checkbox
         const checkboxLabel = document.createElement("label");
         checkboxLabel.classList.add("form-check-label");
         checkboxLabel.textContent = label;
-        checkboxLabel.setAttribute("for", checkbox.id);
```
Comment (plhrtr) [inline canvas_project/static/js/projectSettingsManager.mjs:116]: same as above

File: canvas_project/static/js/picker.mjs [hunk 4]
```diff
@@ -57,17 +55,21 @@ export class Picker {
...
-        );
-        this.#canvas.addEventListener("mousemove", (event) =>
-            this.#onMouseMove(event)
-        );
-        this.#canvas.addEventListener("mouseup", (event) =>
-            this.#onMouseUp(event)
-        );
-
-        // TODO: Event listener for Rectangular selection (not yet implemented)
+        this.#canvas.addEventListener("mousedown", (event) => {
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:58]: the canvas element where the view is rendered is the last child of this.#canvas 
```suggestion
        this.#canvas.children[this.#canvas.children.length -1 ].addEventListener("mousedown", (event) => {
```
To safe the unecessary isMouseOfNav call

File: canvas_project/static/js/picker.mjs [hunk 5]
```diff
@@ -148,8 +150,16 @@ export class Picker {
 
     #onMouseUp(event) {
         // Only calls onClick if it was a real click  and not a drag
-        if (!this.#isDragging && !this.#isTransformDragging) {
+        if (!this.#isDragging) {
             this.#onClick(event);
+        } else if (this.#transformControls.object) {
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:155]: I would take another look at the implementation in of this in the three js editor, as right now the information where the object was and how it was rotated before is gone. But if you want the update via an command you need this information, to undo it.
(You probably also need to add an option to pass this information to the command that gets executed in the according updatePosition() or updateRotation() command)

File: canvas_project/static/js/picker.mjs [hunk 6]
```diff
@@ -170,6 +180,21 @@ export class Picker {
         // Raycast to find the clicked object
         this.#selectedObject = this.#select(this.#mouse, this.#camera);
 
+        // Check if the selected object is movable or rotatable
+        if (this.#selectedObject) {
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:184]: Do you really want to log an error when the object is not rotable or movable, because thats exected behavivour and not a runtime error.

What you need to do is attach the transformcontrols only on when the object supports the current mode

File: canvas_project/static/js/picker.mjs [hunk 7]
```diff
@@ -304,6 +329,21 @@ export class Picker {
             -((position.y - rect.top) / rect.height) * 2 + 1
         );
     }
+
+    // Helper method to check if the mouse is over the nav elements or their children
+    #isMouseOverNavElements(event) {
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:334]: I would remove this and use the above mentioned solution, as if you add another html element to the canvas then this will break

File: canvas_project/static/js/picker.mjs [hunk 8]
```diff
@@ -170,6 +180,21 @@ export class Picker {
...
+            } else if (this.#transformControls.mode === "translate") {
+                if (!this.#selectedObject.isMovable) { 
+                    console.error("selected Object is not movable");
+                    return;
+                }
+            }
+        }
+
         // Update selection (handles shift-key and multi-selection)
         this.#updateSelection(event.shiftKey);
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:199]: I would use the STRG Key to add to the selction, as it the typical key for that

**Review by plhrtr — CHANGES_REQUESTED** [review]
- I would update the current design to include the execution of commands when moving or rotating an object with the transform controls (the updatePosition() and updateRotation() method of each object should execute the update via an command), because right now this is hardly possible (for that you also propably need to update the commands, to allow an oldPosition and oldRotation to be set manually)
- Also currently the transformControls get attached no matter if the object supports the action or not (isRotable, isMovable)
- The setMode method sould allow the mode "move" instead of "single", also when changing the mode the picker you cant select objects anymore
- add the snap to grid functionality when holding shift

File: canvas_project/editor/templates/editor/editor.html (hunk 1 above)
Comment (lydiaklBal) [reply canvas_project/editor/templates/editor/editor.html:311]: I removed it completely because it was only needed for the isMouseOverNavElements(event) method

File: canvas_project/static/js/picker.mjs [hunk 9]
```diff
@@ -148,16 +134,31 @@ export class Picker {
...
+                */
+                this.#itemSelectedEvent();
+            }
         }
     }
 
     /**
      * Handles the click event on the canvas
      */
     #onClick(event) {
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:160]: Declare a type for the event --> event.strgKey isn't valid code but doesn't get recognized because event has an any type

File: canvas_project/static/js/picker.mjs [hunk 10]
```diff
@@ -170,8 +171,25 @@ export class Picker {
...
+                if (!this.#selectedObject.isMovable) {
+                    console.error("selected Object is not movable");
+                    return;
+                }
+            }
+        }
+        */
+
+        // Update selection (handles strg-key and multi-selection)
+        this.#updateSelection(event.strgKey);
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:192]: should be event.ctrlKey

File: canvas_project/static/js/picker.mjs [hunk 11]
```diff
@@ -274,19 +280,25 @@ export class Picker {
         if (this.#selectedObjects.length === 0) {
             this.#transformControls.detach();
         } else if (this.#selectedObjects.length === 1) {
+            if (this.#transformControls.mode === "rotate") {
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:283]: Right now if a object doesn't support for example rotate the selectionBox also doesn't get attached

File: canvas_project/static/js/picker.mjs [hunk 12]
```diff
@@ -274,19 +280,25 @@ export class Picker {
...
+                if (!this.#selectedObject.isRotatable) {
+                    this.#transformControls.detach();
+                    return;
+                }
+                if (this.#selectedObject instanceof Receiver) {
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:288]: Maybe use a solution that doesn't require instance of. Maybe is rotatable return null if not rotatable or a list of axis where it is

File: canvas_project/static/js/objects.mjs [hunk 13]
```diff
@@ -381,11 +433,10 @@ export class Receiver extends SelectableObject {
         super(receiverName);
         // place the 3D object
         this.#base = new ReceiverBase();
-        this.#base.position.copy(new THREE.Vector3(position.x, 0, position.z));
+        this.#base.position.copy(new THREE.Vector3(0, -position.y, 0));
```
Comment (plhrtr) [inline canvas_project/static/js/objects.mjs:436]: The receiver position right now doesn't match the position that is shown and represented in the inspector, if i set a receiver to (0, 50, 0) it is at (0,0,0)

File: canvas_project/static/js/picker.mjs [hunk 14]
```diff
@@ -148,16 +134,31 @@ export class Picker {
 
     #onMouseUp(event) {
         // Only calls onClick if it was a real click  and not a drag
-        if (!this.#isDragging && !this.#isTransformDragging) {
+        if (!this.#isDragging) {
             this.#onClick(event);
+        } else if (this.#transformControls.object) {
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:139]: And when I select a light source the old selectionBox is still attached to the last selected, because it doesn't get detached each time the selection changed

File: canvas_project/static/js/picker.mjs [hunk 15]
```diff
@@ -274,19 +280,25 @@ export class Picker {
         if (this.#selectedObjects.length === 0) {
             this.#transformControls.detach();
         } else if (this.#selectedObjects.length === 1) {
+            if (this.#transformControls.mode === "rotate") {
+                if (!this.#selectedObject.isRotatable) {
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:284]: And when I select a light source the old selectionBox is still attached to the last selected, because it doesn't get detached each time the selection changed

**Review by plhrtr — CHANGES_REQUESTED** [review]
- Maybe add a method to determine if the object is selectable to attache the seleciton box or not, not depending on if the objects is rotable or movable 

Besides the comment I think it looks good and I would merge it when those things get fixed. And for the update of the rotation and position through commands I think it would be best to open another issue

File: canvas_project/static/js/picker.mjs (hunk 11 above)
Comment (lydiaklBal) [reply canvas_project/static/js/picker.mjs:283]: wdym? so you are saying that the box should be attached to the objects even when it is not rotatable?
or do you mean that the boxes need to be deselected too?

File: canvas_project/static/js/picker.mjs (hunk 12 above)
Comment (lydiaklBal) [reply canvas_project/static/js/picker.mjs:288]: wdym? The idea was that when a Receiver is selected and the rotation mode is on, only the y-axis for rotation is shown, so that no other rotations can be applied. 
the instanceof could be completely removed, since the Receiver is the only rotatable object, but idk if this leads to any other issue down the road. (e.g. maybe if new rotateble object types added which require the other 2 axis)

File: canvas_project/static/js/picker.mjs (hunk 11 above)
Comment (lydiaklBal) [reply canvas_project/static/js/picker.mjs:295]: I changed it so that the boxes are deselected too, but I can change it back when needed

Comment (plhrtr) [reply canvas_project/static/js/picker.mjs:295]: > wdym? so you are saying that the box should be attached to the objects even when it is not rotatable? or do you mean that the boxes needs to be deselected too?

Because the selection box highlights the selected item it should be attached no matter what (except objects like the lightsource that aren't really 3D objects), but right now the selection box only gets attached along side the transfrom controls, so if an object isn't rotatable and you click on it with the rotation mode enabled it wont be highlighted

File: canvas_project/static/js/updateCommands.mjs [hunk 16]
```diff
@@ -70,7 +70,10 @@ export class UpdateHeliostatCommand extends SingleObjectCommand {
             case "kinematicType":
                 this.#heliostat.kinematicType = this.#newParameter;
                 break;
-            case "position":
+                case "position":
+                console.log("paremeter",this.#newParameter);
```
Comment (plhrtr) [inline canvas_project/static/js/updateCommands.mjs:74]: please remove the console logs

File: canvas_project/static/js/picker.mjs [hunk 17]
```diff
@@ -146,18 +146,51 @@ export class Picker {
...
+     * @param {MouseEvent} event
+     */
     #onMouseUp(event) {
         // Only calls onClick if it was a real click  and not a drag
-        if (!this.#isDragging && !this.#isTransformDragging) {
+        if (!this.#isDragging) {
             this.#onClick(event);
+        } else if (this.#transformControls.object) {
+            if (this.#transformControls.mode === "translate") {
+                if (this.#selectedObject instanceof Heliostat) {
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:158]: I think it would be better to add a updateAndSavePosition method like the updateAndSaveName method in the Selectable object class and override it in every subclass to prevent the instance of here (makes it more expandable if another class gets added later, because then the picker can handle it by default. For example: For a subclass of a heliostat with more specific features, different commands are most likely needed to handle the new parameters and right now this subclass would break the picker). 

The updateAndSavePosition (and same for rotation) of the lightsource cant be executed because the transform controls never will be attached so for this class the method dont need to be overriden

File: canvas_project/static/js/picker.mjs [hunk 18]
```diff
@@ -146,18 +146,51 @@ export class Picker {
...
+                        new UpdateReceiverCommand(
+                            this.#selectedObject,
+                            "position",
+                            this.#transformControls.object.position.clone()
+                        )
+                    );
+                }
+                this.#itemSelectedEvent();
+            } else if (this.#transformControls.mode === "rotate") {
+                //TODO: auch mit Commands Updaten
```
Comment (plhrtr) [inline canvas_project/static/js/picker.mjs:177]: Do you want to add the Rotate through command in this pull request or later? Because then i will wait with merging till implemented :)

**Review by plhrtr — COMMENTED** [review]
It's really nice to have the transform control changes be executed through commands!!

File: canvas_project/static/js/picker.mjs (hunk 18 above)
Comment (lydiaklBal) [reply canvas_project/static/js/picker.mjs:177]: I would create a new brach for it :)

**Review by plhrtr — CHANGES_REQUESTED** [review]
Looks really nice!

Please consider the comments above, but now only some minor changes are needed till it's finished :) 

Do you want to add the Rotate through command in this pull request or later? Because then i will wait with merging till implemented :)

**Comment by lydiaklBal** [comment]
I will create a new branch to fix the rotation issue

**Comment by plhrtr** [comment]
lgtm!
